specs/merkle-tree.md

```diff
@@ -0,0 +1,142 @@
+* `Enum.Operation operation`: The type of transaction to execute in the Gnosis Safe, i.e. `Call` or `DelegateCall`.
+* `bool requireSuccess`: If this is `true` and the transaction in the Gnosis Safe fails, the deployment is marked as "failed" and will end immediately. If this is `false`, the deployment will continue regardless of whether the transaction fails.
+
+### `CANCEL` Leaf Data
+TODO
+
+## Valid Merkle Trees
+There are a wide variety of trees that conform to the basic structure described above, but not all of them will necessarily be executable on-chain for all networks they contain leaves for. For example, it is possible to include two leaves with the same `index` and `chainId`. In that case, both leaves would be valid but only one would be executable.
+
+To resolve these types of problems we outline a set of Merkle tree invariants below that define what a 'valid' Merkle tree looks like. We expect relayers to rely on Merkle trees following the invariants set out here. We assume that relayers will refuse to execute Merkle trees that do not follow these invariants.
+
+## High-Level Merkle Tree Invariants
+
+### 1. Must be executable on all chains for which there is at least one Merkle leaf.
+Assumption: The input Merkle leaf(s) data satisfies the on-chain conditions in the corresponding SphinxModuleProxy contract(s). See the [SphinxModuleProxy specification](https://github.com/sphinx-labs/sphinx/blob/feature/pre-audit/specs/sphinx-module-proxy.md#function-level-invariants) for more information.
+
+### 2. Each leaf on a chain must have an index field with a value that sequentially increments by 1
+To ensure that `EXECUTE` leaves are executed in the correct order, all Merkle tree leaves must have an `index` field that starts at 0 on each chain and sequentially increments by 1.
+
+### 3. Every leaf `chainId` and `index` combination must be unique within the tree
+Constructing a tree with multiple leaves with the same chain ID and index would make it ambiguous which leaf should be executed, and would result in a tree that cannot be fully executed. So we impose the restriction that in a valid Merkle tree there must be at most one leaf with each chain ID and index combination which ensures there is unambiguous ordering of leaves within the tree.
+
+### 4. There must be exactly one `APPROVE` leaf or `CANCEL` leaf per chain
+Constructing a tree that contains `APPROVE` or `CANCEL` leaves that approve or cancel multiple separate deployments on a single chain would lead to an ambiguous situation for the relayer where it is unclear which deployment should be approved or canceled. So we impose the restriction that in a valid Merkle tree there must be exactly one `APPROVE` or `CANCEL` leaf per chain.
+
+If the `arbitraryChain` field is set to true, then we consider this approval leaf to apply to all chains. Therefore if `arbitraryChain` = true, there should be exactly one `APPROVE` leaf in the entire tree and no `CANCEL` leaves.
```

Comment:
Auditors will check that this is upheld in the TypeScript logic. Do we enforce this?